redis/src/connection/io/mod.rs

```diff
@@ -0,0 +1,277 @@
+use std::{io::Write as _, net, time::Duration};
+
+use crate::{connection::addr::ConnectionAddr, Parser, RedisError, RedisResult, Value};
+
+pub(crate) mod tcp;
+pub use tcp::TcpConnection;
+
+#[cfg(all(feature = "tls-native-tls", not(feature = "tls-rustls")))]
+pub(crate) mod tls_native;
+
+#[cfg(all(feature = "tls-native-tls", not(feature = "tls-rustls")))]
+pub use tls_native::TcpNativeTlsConnection;
+
+#[cfg(feature = "tls-rustls")]
+pub(crate) mod tls_rustls;
+
+#[cfg(feature = "tls-rustls")]
+pub use tls_rustls::{ClientTlsConfig, TcpRustlsConnection, TlsCertificates, TlsConnParams};
+
+#[cfg(not(feature = "tls-rustls"))]
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct TlsConnParams;
+
+#[cfg(unix)]
+pub(crate) mod unix;
+
+#[cfg(unix)]
+pub use unix::UnixConnection;
+
+/// Driver trait that can be implemented
+/// to bring your own IO when using Redis synchronously.
+///
+/// Most users however will probably just use [`ActualConnection`] instead.
+pub trait ConnectionDriver: Sized {
+    /// Establish a connection, returning handle for command execution.
+    fn establish_conn(addr: &ConnectionAddr, timeout: Option<Duration>) -> RedisResult<Self>;
+
+    /// Sets the write timeout for the connection.
+    fn set_write_timeout(&self, dur: Option<Duration>) -> RedisResult<()>;
+    /// Sets the read timeout for the connection.
+    fn set_read_timeout(&self, dur: Option<Duration>) -> RedisResult<()>;
+
+    /// Send bytes over the connection driver.
+    fn send_bytes(&mut self, bytes: &[u8]) -> RedisResult<Value>;
+
+    /// Read a Redis value from the connection driver.
+    fn read_value(&mut self, parser: &mut Parser) -> RedisResult<Value>;
+
+    /// Reports if the connection driver is open.
+    fn is_open(&self) -> bool;
+
+    /// Closes the connection driver.
+    fn close_connection(&mut self);
```

Comment:
why is this needed? wouldn't it just be better to close on drop?

Reply:
Probably yes. I think I simply took it over from the original API and moved it into the trait.
Your comment is valid though. Want me to try to remove this and move it into drop logic?

Reply:
Yes, please. I prefer to make public traits minimal.

Reply:
It's a change from original code though as the original `Connection` used to have already 'close_connection' which now is changed to the below because of your request:

```rust
fn close_connection(&mut self) {
        // Notify the PushManager that the connection was lost
        self.send_disconnect();
        // NOTE: In original implementation this used to call `self.con.close_connection()`
        // now this happens in Drop
    }
```

Might be fine, but just saying that was the reason why I moved it also in the trait, to stay compatible with exact original behaviour